### PR TITLE
Initial support for Apple Silicon

### PR DESCRIPTION
**Pull request by buchwasa**
This PR depends on https://github.com/GitCredentialManager/git-credential-manager/pull/635

This PR aims to add support for Apple Silicon.
When compiling, the build script will now automatically select x64 or arm64 to compile with.
Avalonia has been bumped to version 0.10.13 as 0.10.7 crashed on arm64 (x64 through Rosetta worked fine on that version, however).

I have tested this PR through Rosetta and native and everything works perfectly, however, this is lacking sufficient testing on linux-x64.

### CODE REVIEW

**Comment by sn3akrr** [comment]
bad

File: src/shared/Atlassian.Bitbucket.UI.Avalonia/Atlassian.Bitbucket.UI.Avalonia.csproj
Comment (ldennington) [inline src/shared/Atlassian.Bitbucket.UI.Avalonia/Atlassian.Bitbucket.UI.Avalonia.csproj:6]: Again, let's verify `osx-arm64` is a [supported RID](https://docs.microsoft.com/en-us/dotnet/core/rid-catalog#macos-rids) (ditto for all the other `.csproj` files.

File: src/osx/Installer.Mac/layout.sh
Comment (ldennington) [inline src/osx/Installer.Mac/layout.sh:37]: Verify this is a [supported RID](https://docs.microsoft.com/en-us/dotnet/core/rid-catalog#macos-rids). Also, although it would be ideal to publish a [Universal binary](https://developer.apple.com/documentation/apple-silicon/building-a-universal-macos-binary), dotnet doesn't currently support this (I found a [thread](https://docs.microsoft.com/en-us/answers/questions/471880/net-core-support-in-new-mac-os-arm-based-processor.html) that suggests it _might_ be coming with .NET 7). So we'll probably want to build _and publish_ two applications - one for Intel Macs and one for M1 Macs.

**Review by ldennington — COMMENTED** [review]
This is a great start! I've left a couple comments and also have one high-level note:

We're going to have to update CI to publish packages for both `x86_64` and `ARM64`. The relevant CI/release pipelines can be found in the [`.azure-pipelines` directory](https://github.com/GitCredentialManager/git-credential-manager/tree/main/.azure-pipelines). This could end up being a _bit_ tricky, given that these pipelines run in a restricted Azure DevOps account, but we can work together to validate changes as needed.

File: src/shared/Atlassian.Bitbucket.UI.Avalonia/Atlassian.Bitbucket.UI.Avalonia.csproj
Comment (buchwasa) [reply src/shared/Atlassian.Bitbucket.UI.Avalonia/Atlassian.Bitbucket.UI.Avalonia.csproj:6]: Per the [dotnet runtime repo](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.NETCore.Platforms/src/runtime.json#L2117l), osx-arm64 is a supported RID. I could not verify if it was a supported RID in .NET 5, but it definitely is in .NET 6.

File: src/osx/Installer.Mac/layout.sh
Comment (buchwasa) [reply src/osx/Installer.Mac/layout.sh:37]: See above comment. I was disappointed that universal binaries were not supported yet in dotnet, I'll keep an eye out once this PR gets merged and update that. I did want to add cross-compiling that way those releases can be published like https://github.com/GitCredentialManager/git-credential-manager/pull/237 has, but I decided to hold off until I got some feedback.

File: src/shared/Atlassian.Bitbucket.UI.Avalonia/Atlassian.Bitbucket.UI.Avalonia.csproj
Comment (ldennington) [reply src/shared/Atlassian.Bitbucket.UI.Avalonia/Atlassian.Bitbucket.UI.Avalonia.csproj:6]: Excellent!

**Comment by buchwasa** [comment]
Hello! So sorry that I haven't gotten to this recently, been very busy. Is there anywhere I can contact someone regarding the Azure Pipelines stuff (@ldennington)? I don't want to be spamming this PR asking if it worked or not.

**Comment by ldennington** [comment]
@buchwasa - ignore the check saying that the release build failed. I re-ran with [cf2eec3](https://github.com/GitCredentialManager/git-credential-manager/commit/cf2eec3e06720cd6b632d5bf0a50a4d8e7c9caa4) and it passed. However, it still doesn't look like we're publishing separate packages for `x86_64` and `ARM64`.

Screen Shot 2022-04-04 at 1 04 07 PM

Looking through the commits, I don't think that change has been made yet (forgive me if I missed it). Once this change has been applied, I can re-run the pipeline to verify the separate packages are generated.

**Comment by ldennington** [comment]
@buchwasa - also, while I _definitely_ understand how it feels to have a case of the Mondays ;), I wonder if you wouldn't mind making your commit messages a little more descriptive? One of my co-workers just gave an _excellent_ talk on commit message formatting with some helpful guidance:

https://www.youtube.com/watch?v=4qLtKx9S9a8

If you'd rather just view the slides, you can find them here:

https://www.youtube.com/watch?v=4qLtKx9S9a8

**Comment by buchwasa** [comment]
> @buchwasa - ignore the check saying that the release build failed. I re-ran with [cf2eec3](https://github.com/GitCredentialManager/git-credential-manager/commit/cf2eec3e06720cd6b632d5bf0a50a4d8e7c9caa4) applied and it passed. However, it still doesn't look like we're publishing separate packages for `x86_64` and `ARM64`.
> 
> Screen Shot 2022-04-04 at 1 04 07 PM
> 
> Looking through the commits, I don't think that change has been made yet (forgive me if I missed it). Once this change has been applied, I can re-run the pipeline to verify the separate packages are generated.

At the time, I didn't have any code committed, I just didn't want to spam people who have notifications on for this repo

**Comment by buchwasa** [comment]
This should be ready for another test, I didn't start allowing to be building both, because I want to make sure that the initial changes at least work. Apologies if this stuff is super obvious, I haven't worked with Azure Pipelines yet, so this is a big trial and error thing for me.

**Comment by ldennington** [comment]
> This should be ready for another test, I didn't start allowing to be building both, because I want to make sure that the initial changes at least work. Apologies if this stuff is super obvious, I haven't worked with Azure Pipelines yet, so this is a big trial and error thing for me.

No need to apologize - this is in no way your fault. The current setup with Azure Pipelines is incredibly opaque for our contributors. I am currently working to migrate it to GitHub actions, at which point you will be able to run these types of tests yourself.

**Comment by wendylove14700** [comment]
Netel

**Comment by wendylove14700** [comment]
Octacat

**Comment by wendylove14700** [comment]
Neteller
[](neteller)

**Comment by ldennington** [comment]
Sorry for the delay @buchwasa - there were a few issues with the release build that I worked through. I've [opened a PR](https://github.com/buchwasa/git-credential-manager/pull/1) into this branch with my updates; once those are merged, I think you should be good to go with side-by-side publishing of the macOS binaries!

**Comment by buchwasa** [comment]
> > This should be ready for another test, I didn't start allowing to be building both, because I want to make sure that the initial changes at least work. Apologies if this stuff is super obvious, I haven't worked with Azure Pipelines yet, so this is a big trial and error thing for me.
> 
> No need to apologize - this is in no way your fault. The current setup with Azure Pipelines is incredibly opaque for our contributors. I am currently working to migrate it to GitHub actions, at which point you will be able to run these types of tests yourself. For now, though, I've kicked off a new Azure Pipelines run and will report on results tomorrow!

Oh cool! I've actually worked a lot with GitHub actions in the recent months, and would love to help out with that, if possible!

**Comment by buchwasa** [comment]
Maybe I'm overthinking, but would it be best to just add a separate job in release.yml for osx arm64?

**Comment by ldennington** [comment]
> Maybe I'm overthinking, but would it be best to just add a separate job in release.yml for osx arm64?

You're not overthinking - I agree that this would probably be a good idea at some point. However, my work to migrate our release flows to GitHub Actions led me to discover [this issue](https://github.com/actions/runner/issues/805). It seems that there is no real support for either self-hosted or GitHub-hosted `arm64` runners. In light of this, I actually don't think it's worth trying to create a separate release workflow or build two different binaries at this time.

**Comment by buchwasa** [comment]
Isn't it possible to cross compile for arm64 on x64? I know that I can cross compile for x64 on arm64 without Rosetta, I'd imagine it can go either way

**Comment by ldennington** [comment]
@buchwasa - you are correct, cross compilation is possible. Good call on that. However, we still need to break this out into separate installers for `arm64` and `x64`. This will involve creating a new `distribution.xml` file, perhaps we can have `distribution.arm64.xml` and `distribution.x64.xml`.

**Comment by buchwasa** [comment]
I'm not sure if I can pass a parameter to a template and set it as a variable, otherwise, I'd do that to reduce the amount of parameters I have in build.yml (and there is one more parameter that I plan on adding)

**Comment by ldennington** [comment]
This is looking good! I can kick off another Azure DevOps run for validation. And don't sweat the amount of parameters in `build.yml`. Templates aren't supported with GitHub actions, so that's going to undergo a major refactor anyway.

**Comment by ldennington** [comment]
An update: I spent a good part of the day trying to get this to work in Azure Pipelines but couldn't quite get it there. @mjcheetham - what do you think about waiting for the migration to GitHub actions to land then fast following with some refactoring of this PR to make it work with the new infrastructure? Happy to partner with @buchwasa on that (since he should have more visibility into the new workflows).

**Comment by mjcheetham** [comment]
> @mjcheetham - what do you think about waiting for the migration to GitHub actions to land then fast following with some refactoring of this PR to make it work with the new infrastructure?

That's probably a good call to avoid wasting time getting the old CI working just to throw it away shortly afterwards. Right now native Apple Silicon support is not a blocker as running the x86_64 version under Rosetta works OK.

**Comment by buchwasa** [comment]
> Happy to partner with @buchwasa on that (since he should have more visibility into the new workflows).
Keep me posted on this, I'd be happy to help!

**Review by Pilot61 — APPROVED** [review]

**Comment by buchwasa** [comment]
> Hey @buchwasa! Sorry for the delay, but we've officially migrated our release processes to [GitHub Actions](https://github.com/GitCredentialManager/git-credential-manager/blob/main/.github/workflows/release.yml)! I've done some work to update this branch to work with these changes [in my `m1-release-updates` branch](https://github.com/ldennington/git-credential-manager/tree/m1-release-updates). However, this `m1` branch will need to be rebased onto `main` before I can open a PR into it with my changes. Would you mind doing the rebase when you get a chance?

There ya are. This is like the second time where I've successfully rebased, so I have high hopes that it's smooth sailing from here

**Comment by ldennington** [comment]
> There ya are. This is like the second time where I've successfully rebased, so I have high hopes that it's smooth sailing from here

Great job and thank you! PR is [here](https://github.com/buchwasa/git-credential-manager/pull/2).

**Review by ldennington — APPROVED** [review]

**Pull request merged**
